framework/src/main/java/org/checkerframework/common/purity/PurityVisitor.java

```diff
@@ -0,0 +1,196 @@
+package org.checkerframework.common.purity;
+
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import java.util.*;
+import javax.lang.model.type.TypeKind;
+import org.checkerframework.checker.compilermsgs.qual.CompilerMessageKey;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+import org.checkerframework.dataflow.qual.Pure;
+import org.checkerframework.dataflow.util.PurityChecker;
+import org.checkerframework.dataflow.util.PurityUtils;
+import org.checkerframework.framework.source.Result;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
+import org.checkerframework.javacutil.Pair;
+import org.checkerframework.javacutil.TreeUtils;
+
+public class PurityVisitor extends BaseTypeVisitor<PurityAnnotatedTypeFactory> {
+
+    public PurityVisitor(BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    @Override
+    public Void visitMethod(MethodTree node, Void p) {
+        boolean anyPurityAnnotation = PurityUtils.hasPurityAnnotation(atypeFactory, node);
+        boolean checkPurityAlways = checker.hasOption("suggestPureMethods");
+        boolean checkPurityAnnotations = checker.hasOption("checkPurityAnnotations");
+
+        if (checkPurityAnnotations && (anyPurityAnnotation || checkPurityAlways)) {
+            // check "no" purity
+            List<Pure.Kind> kinds = PurityUtils.getPurityKinds(atypeFactory, node);
+            // @Deterministic makes no sense for a void method or constructor
+            boolean isDeterministic = kinds.contains(Pure.Kind.DETERMINISTIC);
+            if (isDeterministic) {
+                if (TreeUtils.isConstructor(node)) {
+                    checker.report(Result.warning("purity.deterministic.constructor"), node);
+                } else if (TreeUtils.typeOf(node.getReturnType()).getKind() == TypeKind.VOID) {
+                    checker.report(Result.warning("purity.deterministic.void.method"), node);
+                }
+            }
+
+            // Report errors if necessary.
+            PurityChecker.PurityResult r =
+                    PurityChecker.checkPurity(
```

Comment:
What method is this calling? I don't see a `checkPurity` method in `PurityChecker`, nor a `PurityResult` class. Can you explain what's going on here?

Reply:
The `PurityChecker` class that it refers to is `org.checkerframework.dataflow.util.PurityChecker`. I agree that having two `PurityChecker` classes is confusing, but it was the initial design to include purity checking in the dataflow folder, so I haven't changed it.

Reply:
I suggest at least renaming that class to avoid the collision. I would suggest calling it `PurityVisitor`, but that would be no better in terms of name collisions.

A better solution would be to move all of the code in that class to `PurityVisitor`, and delete that class.